drivers: adc: fix ADC_DT_SPEC* helpers for channels >= 10

aurel32 opened the pull request:
There might be better way to fix the issue, but it is the only one I have found so far

galak commented:
> There might be better way to fix the issue, but it is the only one I have found so far

what's the issue will values > 10?

aurel32 commented:
> > There might be better way to fix the issue, but it is the only one I have found so far
> 
> what's the issue will values > 10?

Sorry I included the link to the bug report in the commit, but not in the PR description, here it is: https://github.com/zephyrproject-rtos/zephyr/issues/47119

In short, configuration for channels >= 10 are ignored.

galak committed:
sys: util: add a Z_IS_EQ macro to compare integers

This macro compares two integers from 0 to 255 at compile time, using
the preprocessor.

Signed-off-by: Kumar Gala <galak@kernel.org>
Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

galak reviewed (changes requested):
I think we can do this better with some additional macrobatics.

The idea is adding an integer comparison macro.  We can than walk all the children of the ADC controller looking for a matching reg and return that node.

I've got this kinda working, but what I can't figure it is the magic around `IF_ENABLED` and `__DEBRACKET`.

If I do:

`IF_ENABLED(1, foobar)`

I get:

`__DEBRACKET foobar`

I'd expect to get just `foobar`.

@carlescufi @jakub-uC do you know why that is the case?

galak commented:
Here's the `Z_IS_EQ()`

https://github.com/galak/zephyr/commit/94eaddf7c436919851e37da55a61c4650de8069e

galak commented:
> I've got this kinda working, but what I can't figure it is the magic around `IF_ENABLED` and `__DEBRACKET`.
> 
> If I do:
> 
> `IF_ENABLED(1, foobar)`
> 
> I get:
> 
> `__DEBRACKET foobar`
> 
> I'd expect to get just `foobar`.
> 
> @carlescufi @jakub-uC do you know why that is the case?

I think I figured that part out, was needing an extra `( .. )`

galak commented:
@aurel32 here's my branch that I think is working now:

https://github.com/galak/zephyr/tree/wip/adc-dt-fix

Needs cleaning up, adding of tests, etc..

aurel32 commented:
Thanks @galak, that look way cleaner than my version, and even cleaner than the original version which assumed many things about the naming of the node.

I have tested it on a case with multiple channels, with a mix of channels < 10 and >= 10, and the resulting binary is identical to the one with my solution, so your code works perfectly.

How should we proceed from now? I guess I should just close the PR?

galak commented:
> Thanks @galak, that look way cleaner than my version, and even cleaner than the original version which assumed many things about the naming of the node.
> 
> I have tested it on a case with multiple channels, with a mix of channels < 10 and >= 10, and the resulting binary is identical to the one with my solution, so your code works perfectly.
> 
> How should we proceed from now? I guess I should just close the PR?

Do you mind picking up my code and cleaning it up, adding some tests, etc.

aurel32 commented:
> > Thanks @galak, that look way cleaner than my version, and even cleaner than the original version which assumed many things about the naming of the node.
> > I have tested it on a case with multiple channels, with a mix of channels < 10 and >= 10, and the resulting binary is identical to the one with my solution, so your code works perfectly.
> > How should we proceed from now? I guess I should just close the PR?
> 
> Do you mind picking up my code and cleaning it up, adding some tests, etc.

Yes, I can definitely pick-up your code and clean it. That said I am not sure I fully understand all the macro stuff, so I am not really sure how to test them, at least not without a bit of help.

galak commented:
> Yes, I can definitely pick-up your code and clean it. That said I am not sure I fully understand all the macro stuff, so I am not really sure how to test them, at least not without a bit of help.

Sure, happy to help on that front.

aurel32 committed:
sys: util: export Z_IS_EQ macro as IS_EQ

And also add the corresponding documentation.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
tests: unit: util: Add test for IS_EQ

Add a basic test for the just introduced IS_EQ macro.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
drivers: adc: fix ADC_DT_SPEC* helpers for channels >= 10

Use the just introduced IS_EQ() to get the channel child node matching the
input.

Fixes #47119

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
> > Yes, I can definitely pick-up your code and clean it. That said I am not sure I fully understand all the macro stuff, so I am not really sure how to test them, at least not without a bit of help.
> 
> Sure, happy to help on that front.

@galak, I have pushed a version based on your `Z_IS_EQ` macro. However it seems that your code is not compliant with the coding style, and I am not sure how to fix that:

```
-:50: ERROR:SPACING: space prohibited before that ',' (ctx:WxE)
#50: FILE: include/zephyr/sys/util_internal.h:150:
+#define Z_IS_0_EQ_0(...) ,
                          ^
```

galak commented:
> > 
> 
> @galak, I have pushed a version based on your `Z_IS_EQ` macro. However it seems that your code is not compliant with the coding style, and I am not sure how to fix that:
> 
> ```
> -:50: ERROR:SPACING: space prohibited before that ',' (ctx:WxE)
> #50: FILE: include/zephyr/sys/util_internal.h:150:
> +#define Z_IS_0_EQ_0(...) ,
>                           ^
> ```

I'm guessing this might be a case we request that we avoid the coding style.

jeremybettis commented on include/zephyr/sys/util_internal.h:
```diff
@@ -49,6 +49,8 @@
  */
 #define Z_IS_ENABLED3(ignore_this, val, ...) val
 
+#define Z_IS_EQ(_0, _1) Z_HAS_COMMA(Z_CAT4(Z_IS_, _0, _EQ_, _1)())
```
Seems like a comment will be helpful. Something like.

/* Implementation of IS_EQ(). Returns 1 if _0 and _1 are the same integer from 0 to 255, 0 otherwise. */

aurel32 replied:
Thanks for the suggestion, fixed.

jeremybettis commented on include/zephyr/sys/util_macro.h:
```diff
@@ -229,6 +229,15 @@ extern "C" {
+ * @brief Like <tt>a == b</tt>, but does evaluation and
+ * short-circuiting at C preprocessor time.
+ *
+ * This is however only works for integer literal from 0 to 255.
```
Remove "is"

aurel32 replied:
Thanks for the fix.

aurel32 commented:
> A slight hack, but would putting in , instead of , make checkpatch happy?

What do you mean exactly? For me it seems the same character, `,`

Note also that there are two other macros in that file having the same issue:

```
/tmp/test.patch:84: ERROR:SPACING: space prohibited before that ',' (ctx:WxE)
#84: FILE: include/zephyr/sys/util_internal.h:78:
+#define Z_TRIGGER_PARENTHESIS_(...) ,
                                     ^
/tmp/test.patch:95: ERROR:SPACING: space prohibited before that ',' (ctx:WxE)
#95: FILE: include/zephyr/sys/util_internal.h:89:
+#define Z_IS_EMPTY_CASE_0001 ,

```

jeremybettis commented:
> > A slight hack, but would putting in , instead of , make checkpatch happy?
> 
> What do you mean exactly? For me it seems the same character, `,`
> 
> Note also that there are two other macros in that file having the same issue:
> 
> ```
> /tmp/test.patch:84: ERROR:SPACING: space prohibited before that ',' (ctx:WxE)
> #84: FILE: include/zephyr/sys/util_internal.h:78:
> +#define Z_TRIGGER_PARENTHESIS_(...) ,
>                                      ^
> /tmp/test.patch:95: ERROR:SPACING: space prohibited before that ',' (ctx:WxE)
> #95: FILE: include/zephyr/sys/util_internal.h:89:
> +#define Z_IS_EMPTY_CASE_0001 ,
> ```

Sorry github ate my comment. That was BACKSLASH COMMA. I.e. `#define Z_TRIGGER_PARENTHESIS_(...) \,`

aurel32 commented:
> > > A slight hack, but would putting in , instead of , make checkpatch happy?
> > 
> > 
> > What do you mean exactly? For me it seems the same character, `,`
> > Note also that there are two other macros in that file having the same issue:
> > ```
> > /tmp/test.patch:84: ERROR:SPACING: space prohibited before that ',' (ctx:WxE)
> > #84: FILE: include/zephyr/sys/util_internal.h:78:
> > +#define Z_TRIGGER_PARENTHESIS_(...) ,
> >                                      ^
> > /tmp/test.patch:95: ERROR:SPACING: space prohibited before that ',' (ctx:WxE)
> > #95: FILE: include/zephyr/sys/util_internal.h:89:
> > +#define Z_IS_EMPTY_CASE_0001 ,
> > ```
> 
> Sorry github ate my comment. That was BACKSLASH COMMA. I.e. `#define Z_TRIGGER_PARENTHESIS_(...) \,`

Thanks, that did the trick! I have just pushed a fixed version.
